<?xml version="1.0" ?>
<robot name="fr3_xls">
  <link name="world"/>

  <!-- Virtual -->
  <joint name="world_to_vx" type="fixed">
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <parent link="world"/>
    <child  link="v_x_link"/>
  </joint>
  <link name="v_x_link" />
  <joint name="v_x_joint" type="prismatic">
    <origin rpy="0 0 0" xyz="0 0 0"/>
      <parent link="v_x_link"/>
      <child link="v_y_link"/>
      <axis xyz= "1 0 0"/>
      <limit effort="1000.0" lower="-10000.0" upper="10000.0" velocity="10000.0"/>
  </joint>
  <link name="v_y_link" />
  <joint name="v_y_joint" type="prismatic">
    <origin rpy="0 0 0" xyz="0 0 0"/>
      <parent link="v_y_link"/>
      <child link="v_t_link"/>
      <axis xyz= "0 1 0"/>
      <limit effort="1000.0" lower="-10000.0" upper="10000.0" velocity="10000.0"/>
  </joint>
  <link name="v_t_link" />
  <joint name="v_t_joint" type="revolute">
    <origin rpy="0 0 0" xyz="0 0 0"/>
      <parent link="v_t_link"/>
      <child link="base_footprint"/>
      <axis rpy="0 0 0" xyz="0 0 1"/>
      <limit effort="1000.0" lower="-6.24" upper="6.24" velocity="10000.0"/>
  </joint>

  <!-- Mobile Base -->
  <link name="base_footprint"/>
  <joint name="base_footprint_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0.127"/>
    <parent link="base_footprint"/>
    <child link="base_link"/>
  </joint>

  <link name="base_link">
    <inertial>
      <mass value="125.0"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="1.391" ixy="0.004" ixz="0.0" iyy="6.853" iyz="0.0" izz="6.125"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://mujoco_ros_sim/mujoco_menagerie/fr3_xls/assets/xls/bases/xls/summit_xls_chassis.stl"/>
      </geometry>
      <material name="black">
        <color rgba="0.0 0.0 0.0 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0 " xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://mujoco_ros_sim/mujoco_menagerie/fr3_xls/assets/xls/bases/xls/summit_xls_chassis.stl"/>
      </geometry>
    </collision>
  </link>

  <link name="base_chapa">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://mujoco_ros_sim/mujoco_menagerie/fr3_xls/assets/xls/bases/xls/summit_xls_chapas_inox_tapas.stl"/>
      </geometry>
      <material name="darkgrey">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
  </link>

  <joint name="base_tapa_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0.0 0 0.38062"/>
    <parent link="base_link"/>
    <child link="base_chapa"/>
  </joint>

  <link name="base_logo_left">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://mujoco_ros_sim/mujoco_menagerie/fr3_xls/assets/xls/bases/xls/robotnik_logo_chasis.stl"/>
      </geometry>
      <material name="darkgrey">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
  </link>

  <joint name="base_logo_left_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0.0 0.28931 0.29062"/>
    <parent link="base_link"/>
    <child link="base_logo_left"/>
  </joint>

  <link name="base_logo_right">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://mujoco_ros_sim/mujoco_menagerie/fr3_xls/assets/xls/bases/xls/robotnik_logo_chasis.stl"/>
      </geometry>
      <material name="darkgrey">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
  </link>
  <joint name="base_logo_right_joint" type="fixed">
    <origin rpy="0 0 3.141592653589793" xyz="0.0 -0.28931 0.29062"/>
    <parent link="base_link"/>
    <child link="base_logo_right"/>
  </joint>

  <link name="base_logo_front">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://mujoco_ros_sim/mujoco_menagerie/fr3_xls/assets/xls/bases/xls/robotnik_logo_chasis.stl"/>
      </geometry>
      <material name="darkgrey">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
  </link>
  <joint name="base_logo_front_joint" type="fixed">
    <origin rpy="0 0 -1.5707963267948966" xyz="0.345 0.0 0.0"/>
    <parent link="base_link"/>
    <child link="base_logo_front"/>
  </joint>

  <link name="base_logo_rear">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://mujoco_ros_sim/mujoco_menagerie/fr3_xls/assets/xls/bases/xls/robotnik_logo_chasis.stl"/>
      </geometry>
      <material name="darkgrey">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
  </link>
  <joint name="base_logo_rear_joint" type="fixed">
    <origin rpy="0 0 1.5707963267948966" xyz="-0.345 0.0 0.0"/>
    <parent link="base_link"/>
    <child link="base_logo_rear"/>
  </joint>

  <link name="base_docking_contact">
    </link>
  <joint name="base_docking_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0.35 0 0.09"/>
    <parent link="base_link"/>
    <child link="base_docking_contact"/>
  </joint>

  <!-- Pinocchio takes 'continuous' type as nv=2 -->
  <joint name="front_right_wheel" type="revolute">
    <parent link="base_link"/>
    <child link="front_right_wheel_link"/>
    <origin rpy="0 0 0" xyz="0.2225 -0.2045 0.0"/>
    <axis rpy="0 0 0" xyz="0 1 0"/>
    <limit effort="100" velocity="100"/>
    <joint_properties damping="0.0" friction="0.0"/>
  </joint>

  <link name="front_right_wheel_link">
  <visual>
    <geometry>
        <mesh filename="package://mujoco_ros_sim/mujoco_menagerie/fr3_xls/assets/xls/wheels/omni_wheel_2.dae"/>
      </geometry>
      <material name="darkgrey">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="1.5707963267948966 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.087" radius="0.127"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="6.5"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.0303095" ixy="0" ixz="0" iyy="0.05241925" iyz="0" izz="0.0303095"/>
    </inertial>
  </link>

  <joint name="front_left_wheel" type="revolute">
    <parent link="base_link"/>
    <child link="front_left_wheel_link"/>
    <origin rpy="0 0 0" xyz="0.2225 0.2045 0.0"/>
    <axis rpy="0 0 0" xyz="0 1 0"/>
    <limit effort="100" velocity="100"/>
    <joint_properties damping="0.0" friction="0.0"/>
  </joint>

  <link name="front_left_wheel_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://mujoco_ros_sim/mujoco_menagerie/fr3_xls/assets/xls/wheels/omni_wheel_1.dae"/>
      </geometry>
      <material name="darkgrey">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="1.5707963267948966 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.087" radius="0.127"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="6.5"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.0303095" ixy="0" ixz="0" iyy="0.05241925" iyz="0" izz="0.0303095"/>
    </inertial>
  </link>

  <joint name="rear_left_wheel" type="revolute">
    <parent link="base_link"/>
    <child link="rear_left_wheel_link"/>
    <origin rpy="0 0 0" xyz="-0.2225 0.2045 0.0"/>
    <axis rpy="0 0 0" xyz="0 1 0"/>
    <limit effort="100" velocity="100"/>
    <joint_properties damping="0.0" friction="0.0"/>
  </joint>

  <link name="rear_left_wheel_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://mujoco_ros_sim/mujoco_menagerie/fr3_xls/assets/xls/wheels/omni_wheel_1.dae"/>
      </geometry>
      <material name="darkgrey">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="1.5707963267948966 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.087" radius="0.127"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="6.5"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.0303095" ixy="0" ixz="0" iyy="0.05241925" iyz="0" izz="0.0303095"/>
    </inertial>
  </link>

  <joint name="rear_right_wheel" type="revolute">
    <parent link="base_link"/>
    <child link="rear_right_wheel_link"/>
    <origin rpy="0 0 0" xyz="-0.2225 -0.2045 0.0"/>
    <axis rpy="0 0 0" xyz="0 1 0"/>
    <limit effort="100" velocity="100"/>
    <joint_properties damping="0.0" friction="0.0"/>
  </joint>
  
  <link name="rear_right_wheel_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://mujoco_ros_sim/mujoco_menagerie/fr3_xls/assets/xls/wheels/omni_wheel_2.dae"/>
      </geometry>
      <material name="darkgrey">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="1.5707963267948966 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.087" radius="0.127"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="6.5"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.0303095" ixy="0" ixz="0" iyy="0.05241925" iyz="0" izz="0.0303095"/>
    </inertial>
  </link>

  <!-- Manipulator -->
  <joint name="fr3_base_joint" type="fixed">
    <parent link="base_link"/>
    <child link="fr3_link0"/>
    <origin rpy="0 0 0" xyz="0.2 0.0 0.383"/>
  </joint>

  <link name="fr3_link0">
    <visual name="fr3_link0_visual">
      <geometry>
        <mesh filename="package://mujoco_ros_sim/mujoco_menagerie/fr3_xls/assets/fr3/link0.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.075 0 0.06" rpy="0 1.570796 0"/>
      <geometry>
        <cylinder radius="0.07" length="0.03" />
      </geometry>
    </collision>
    <collision>
      <origin xyz="0.02 0 0.03" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.07" length="0.06" />
      </geometry>
    </collision>
    <collision>
      <origin xyz="-0.06 0 0.06" rpy="0 0 0"/>
      <geometry>
        <sphere radius="0.07" />
      </geometry>
    </collision>
    <collision>
      <origin xyz="-0.09 0 0.06" rpy="0 0 0"/>
      <geometry>
        <sphere radius="0.07" />
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="-0.0172 0.0004 0.0745"/>
      <mass value="2.3966"/>
      <inertia ixx="0.009" ixy="0.0" ixz="0.002" iyy="0.0115" iyz="0.0" izz="0.0085"/>
    </inertial>
  </link>

  <link name="fr3_link1">
    <visual name="fr3_link1_visual">
      <geometry>
        <mesh filename="package://mujoco_ros_sim/mujoco_menagerie/fr3_xls/assets/fr3/link1.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 -0.1915" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.07" length="0.2830" />
      </geometry>
    </collision>
    <collision>
      <origin xyz="0 0 -0.333" rpy="0 0 0"/>
      <geometry>
        <sphere radius="0.07" />
      </geometry>
    </collision>
    <collision>
      <origin xyz="0 0 -0.05" rpy="0 0 0"/>
      <geometry>
        <sphere radius="0.07" />
      </geometry>
    </collision>
    <collision>
      <origin xyz="0 -0.05 -0.05" rpy="0.523598667 0 0"/>
      <geometry>
        <cylinder radius="0.065" length="0.12" />
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0.0000004128 -0.0181251324 -0.0386035970"/>
      <mass value="2.9274653454"/>
      <inertia ixx="0.023927316485107913" ixy="1.3317903455714081e-05" ixz="-0.00011404774918616684" iyy="0.0224821613275756" iyz="-0.0019950320628240115" izz="0.006350098258530016"/>
    </inertial>
  </link>

  <joint name="fr3_joint1" type="revolute">
    <origin rpy="0 0 0" xyz="0 0 0.333"/>
    <parent link="fr3_link0"/>
    <child link="fr3_link1"/>
    <axis xyz="0 0 1"/>
    <limit effort="87.0" lower="-2.7437" upper="2.7437" velocity="2.62"/>
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.7437" soft_upper_limit="2.7437"/>
    <dynamics D="1" K="7000" damping="0.003" friction="0.2" mu_coulomb="0" mu_viscous="16"/>
  </joint>

  <link name="fr3_link2">
    <visual name="fr3_link2_visual">
      <geometry>
        <mesh filename="package://mujoco_ros_sim/mujoco_menagerie/fr3_xls/assets/fr3/link2.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 -0.05 0.05" rpy="2.181661111 0 0"/>
      <geometry>
        <cylinder radius="0.065" length="0.12" />
      </geometry>
    </collision>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.07" length="0.12" />
      </geometry>
    </collision>
    <collision>
      <origin xyz="0 0 -0.08" rpy="0 0 0"/>
      <geometry>
        <sphere radius="0.07" />
      </geometry>
    </collision>
    <collision>
      <origin xyz="0 0 0.08" rpy="0 0 0"/>
      <geometry>
        <sphere radius="0.07" />
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0.0031828864 -0.0743221644 0.0088146084"/>
      <mass value="2.9355370338"/>
      <inertia ixx="0.041938946257609425" ixy="0.00020257331521090626" ixz="0.004077784227179924" iyy="0.02514514885014724" iyz="-0.0042252158006570156" izz="0.06170214472888839"/>
    </inertial>
  </link>

  <joint name="fr3_joint2" type="revolute">
    <origin rpy="-1.570796326794897 0 0" xyz="0 0 0"/>
    <parent link="fr3_link1"/>
    <child link="fr3_link2"/>
    <axis xyz="0 0 1"/>
    <limit effort="87.0" lower="-1.7837" upper="1.7837" velocity="2.62"/>
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-1.7837" soft_upper_limit="1.7837"/>
    <dynamics D="1" K="7000" damping="0.003" friction="0.2" mu_coulomb="0" mu_viscous="16"/>
  </joint>

  <link name="fr3_link3">
    <visual name="fr3_link3_visual">
      <geometry>
        <mesh filename="package://mujoco_ros_sim/mujoco_menagerie/fr3_xls/assets/fr3/link3.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 -0.145" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.07" length="0.15" />
      </geometry>
    </collision>
    <collision>
      <origin xyz="0 0 -0.22" rpy="0 0 0"/>
      <geometry>
        <sphere radius="0.07" />
      </geometry>
    </collision>
    <collision>
      <origin xyz="0 0 -0.07" rpy="0 0 0"/>
      <geometry>
        <sphere radius="0.07" />
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0.0407015686 -0.0048200565 -0.0289730823"/>
      <mass value="2.2449013699"/>
      <inertia ixx="0.02410142547240885" ixy="0.002404694559042109" ixz="-0.002856269270114313" iyy="0.01974053266708178" iyz="-0.002104212683891874" izz="0.019044494482244823"/>
    </inertial>
  </link>

  <joint name="fr3_joint3" type="revolute">
    <origin rpy="1.570796326794897 0 0" xyz="0 -0.316 0"/>
    <parent link="fr3_link2"/>
    <child link="fr3_link3"/>
    <axis xyz="0 0 1"/>
    <limit effort="87.0" lower="-2.9007" upper="2.9007" velocity="2.62"/>
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.9007" soft_upper_limit="2.9007"/>
    <dynamics D="1" K="7000" damping="0.003" friction="0.2" mu_coulomb="0" mu_viscous="16"/>
  </joint>

  <link name="fr3_link4">
    <visual name="fr3_link4_visual">
      <geometry>
        <mesh filename="package://mujoco_ros_sim/mujoco_menagerie/fr3_xls/assets/fr3/link4.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.07" length="0.12" />
      </geometry>
    </collision>
    <collision>
      <origin xyz="0 0 0.06" rpy="0 0 0"/>
      <geometry>
        <sphere radius="0.07" />
      </geometry>
    </collision>
    <collision>
      <origin xyz="-0.01 0.01 0.07" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.07" length="0.02" />
      </geometry>
    </collision>
    <collision>
      <origin xyz="0.01 -0.01 -0.07" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.07" length="0.02" />
      </geometry>
    </collision>
    <collision>
      <origin xyz="0 0 -0.06" rpy="0 0 0"/>
      <geometry>
        <sphere radius="0.07" />
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="-0.0459100965 0.0630492960 -0.0085187868"/>
      <mass value="2.6155955791"/>
      <inertia ixx="0.03452998321913202" ixy="0.01322552265982813" ixz="0.01015142998484113" iyy="0.028881621933049058" iyz="-0.0009762833870704552" izz="0.04125471171146641"/>
    </inertial>
  </link>

  <joint name="fr3_joint4" type="revolute">
    <origin rpy="1.570796326794897 0 0" xyz="0.0825 0 0"/>
    <parent link="fr3_link3"/>
    <child link="fr3_link4"/>
    <axis xyz="0 0 1"/>
    <limit effort="87.0" lower="-3.0421" upper="-0.1518" velocity="2.62"/>
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-3.0421" soft_upper_limit="-0.1518"/>
    <dynamics D="1" K="7000" damping="0.003" friction="0.2" mu_coulomb="0" mu_viscous="16"/>
  </joint>

  <link name="fr3_link5">
    <visual name="fr3_link5_visual">
      <geometry>
        <mesh filename="package://mujoco_ros_sim/mujoco_menagerie/fr3_xls/assets/fr3/link5.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 -0.26" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.07" length="0.1" />
      </geometry>
    </collision>
    <collision>
      <origin xyz="0 0 -0.32" rpy="0 0 0"/>
      <geometry>
        <sphere radius="0.07" />
      </geometry>
    </collision>
    <collision>
      <origin xyz="0 0 -0.22" rpy="0 0 0"/>
      <geometry>
        <sphere radius="0.07" />
      </geometry>
    </collision>
    <collision>
      <origin xyz="0 0.08 -0.12" rpy="-0.139626311 0 0"/>
      <geometry>
        <cylinder radius="0.045" length="0.17" />
      </geometry>
    </collision>
    <collision>
      <origin xyz="0 0.085 0" rpy="0 0 0"/>
      <geometry>
        <sphere radius="0.07" />
      </geometry>
    </collision>
    <collision>
      <origin xyz="0 0.05 -0.19" rpy="0 0 0"/>
      <geometry>
        <sphere radius="0.06" />
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="-0.0016039605 0.0292536262 -0.0972965990"/>
      <mass value="2.3271207594"/>
      <inertia ixx="0.051610278463662895" ixy="-0.005715173387783472" ixz="-0.0035673167625872135" iyy="0.04787729713371481" iyz="0.010673985108535986" izz="0.016423625579357254"/>
    </inertial>

  </link>
  <joint name="fr3_joint5" type="revolute">
    <origin rpy="-1.570796326794897 0 0" xyz="-0.0825 0.384 0"/>
    <parent link="fr3_link4"/>
    <child link="fr3_link5"/>
    <axis xyz="0 0 1"/>
    <limit effort="12.0" lower="-2.8065" upper="2.8065" velocity="5.26"/>
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.8065" soft_upper_limit="2.8065"/>
    <dynamics D="1" K="7000" damping="0.003" friction="0.2" mu_coulomb="0" mu_viscous="16"/>
  </joint>
  <link name="fr3_link6">
    <visual name="fr3_link6_visual">
      <geometry>
        <mesh filename="package://mujoco_ros_sim/mujoco_menagerie/fr3_xls/assets/fr3/link6.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 -0.03" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.07" length="0.08" />
      </geometry>
    </collision>
    <collision>
      <origin xyz="0.02 0 0.01" rpy="0 1.570796 0"/>
      <geometry>
        <cylinder radius="0.07" length="0.07" />
      </geometry>
    </collision>
    <collision>
      <origin xyz="0 0 0.01" rpy="0 0 0"/>
      <geometry>
        <sphere radius="0.07" />
      </geometry>
    </collision>
    <collision>
      <origin xyz="0 0 -0.08" rpy="0 0 0"/>
      <geometry>
        <sphere radius="0.061" />
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0.0597131221 -0.0410294666 -0.0101692726"/>
      <mass value="1.8170376524"/>
      <inertia ixx="0.005412333594383447" ixy="0.006193456360285834" ixz="0.0014219289306117652" iyy="0.014058329545509979" iyz="-0.0013140753741120031" izz="0.016080817924212554"/>
    </inertial>
  </link>

  <joint name="fr3_joint6" type="revolute">
    <origin rpy="1.570796326794897 0 0" xyz="0 0 0"/>
    <parent link="fr3_link5"/>
    <child link="fr3_link6"/>
    <axis xyz="0 0 1"/>
    <limit effort="12.0" lower="0.5445" upper="4.5169" velocity="4.18"/>
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="0.5445" soft_upper_limit="4.5169"/>
    <dynamics D="1" K="7000" damping="0.003" friction="0.2" mu_coulomb="0" mu_viscous="16"/>
  </joint>

  <link name="fr3_link7">
    <visual name="fr3_link7_visual">
      <geometry>
        <mesh filename="package://mujoco_ros_sim/mujoco_menagerie/fr3_xls/assets/fr3/link7.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0.01" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.055" length="0.14" />
      </geometry>
    </collision>
    <collision>
      <origin xyz="0 0 0.08" rpy="0 0 0"/>
      <geometry>
        <sphere radius="0.05"  />
      </geometry>
    </collision>
    <collision>
      <origin xyz="0 0 -0.06" rpy="0 0 0"/>
      <geometry>
        <sphere radius="0.055"  />
      </geometry>
    </collision>
    <collision>
      <origin xyz="0.04 0.04 0.08" rpy="0 0 0.785398"/>
      <geometry>
        <box size="0.071 0.071 0.041" />
      </geometry>
    </collision>
    <collision>
      <origin xyz="0 0 0.079" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.051"  length="0.05"  />
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0.0045225817 0.0086261921 -0.0161633251"/>
      <mass value="0.6271432862"/>
      <inertia ixx="0.00021092389150104718" ixy="-2.433299114461931e-05" ixz="4.564480393778983e-05" iyy="0.00017718568002411474" iyz="8.744070223226438e-05" izz="5.993190599659971e-05"/>
    </inertial>
  </link>

  <joint name="fr3_joint7" type="revolute">
    <origin rpy="1.570796326794897 0 0" xyz="0.088 0 0"/>
    <parent link="fr3_link6"/>
    <child link="fr3_link7"/>
    <axis xyz="0 0 1"/>
    <limit effort="12.0" lower="-3.0159" upper="3.0159" velocity="5.26"/>
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-3.0159" soft_upper_limit="3.0159"/>
    <dynamics D="1" K="7000" damping="0.003" friction="0.2" mu_coulomb="0" mu_viscous="16"/>
  </joint>

  <link name="fr3_link8"/>

  <joint name="fr3_joint8" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0.107"/>
    <parent link="fr3_link7"/>
    <child link="fr3_link8"/>
  </joint>
  
</robot>
